crates/rollup-boost/src/flashblocks/service.rs

```diff
@@ -0,0 +1,318 @@
+use super::outbound::WebSocketPublisher;
+use super::primitives::{
+    ExecutionPayloadBaseV1, ExecutionPayloadFlashblockDeltaV1, FlashblocksPayloadV1,
+};
+use crate::RpcClientError;
+use crate::{
+    ClientResult, EngineApiExt, NewPayload, OpExecutionPayloadEnvelope, PayloadVersion, RpcClient,
+};
+use alloy_primitives::U256;
+use alloy_rpc_types_engine::{
+    BlobsBundleV1, ExecutionPayloadV1, ExecutionPayloadV2, ExecutionPayloadV3,
+};
+use alloy_rpc_types_engine::{ForkchoiceState, ForkchoiceUpdated, PayloadId, PayloadStatus};
+use alloy_rpc_types_eth::{Block, BlockNumberOrTag};
+use jsonrpsee::core::async_trait;
+use op_alloy_rpc_types_engine::{
+    OpExecutionPayloadEnvelopeV3, OpExecutionPayloadEnvelopeV4, OpExecutionPayloadV4,
+    OpPayloadAttributes,
+};
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::sync::RwLock;
+use tokio::sync::mpsc;
+use tracing::error;
+
+#[derive(Debug, Error)]
+pub enum FlashblocksError {
+    #[error("Missing base payload for initial flashblock")]
+    MissingBasePayload,
+    #[error("Unexpected base payload for non-initial flashblock")]
+    UnexpectedBasePayload,
+    #[error("Missing delta for flashblock")]
+    MissingDelta,
+    #[error("Invalid index for flashblock")]
+    InvalidIndex,
+    #[error("Missing payload")]
+    MissingPayload,
+}
+
+impl From<FlashblocksError> for RpcClientError {
+    fn from(err: FlashblocksError) -> Self {
+        RpcClientError::InvalidPayload(err.to_string())
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct FlashbotsMessage {
+    method: String,
+    params: serde_json::Value,
+    #[serde(default)]
+    id: Option<u64>,
+}
+
+// Simplify actor messages to just handle shutdown
+#[derive(Debug)]
+enum FlashblocksEngineMessage {
+    FlashblocksPayloadV1(FlashblocksPayloadV1),
+}
+
+#[derive(Debug, Default)]
+struct FlashblockBuilder {
+    base: Option<ExecutionPayloadBaseV1>,
+    flashblocks: Vec<ExecutionPayloadFlashblockDeltaV1>,
+}
+
+impl FlashblockBuilder {
+    pub fn new() -> Self {
+        Self {
+            base: None,
+            flashblocks: Vec::new(),
+        }
+    }
+
+    pub fn extend(&mut self, payload: FlashblocksPayloadV1) -> Result<(), FlashblocksError> {
+        tracing::debug!(message = "Extending payload", payload_id = %payload.payload_id, index = payload.index, has_base=payload.base.is_some());
+
+        // Check base payload rules
+        match (payload.index, payload.base) {
+            // First payload must have a base
+            (0, None) => return Err(FlashblocksError::MissingBasePayload),
+            (0, Some(base)) => self.base = Some(base),
+            // Subsequent payloads must have no base
+            (_, Some(_)) => return Err(FlashblocksError::UnexpectedBasePayload),
+            (_, None) => {} // Non-zero index without base is fine
+        }
```

Comment:
Nit: consider simplifying to something like the following:

```suggestion
if payload.index == 0 {
    // First payload must include a base
    if let Some(base) = payload.base {
        self.base = Some(base);
    } else {
        return Err(FlashblocksError::MissingBasePayload);
    }
} else if payload.base.is_some() {
    // Non-zero index must not include a base
    return Err(FlashblocksError::UnexpectedBasePayload);
}
```